code/UI/Interactions/InteractionMenu.cs

```diff
@@ -28,65 +61,95 @@ public void SetEntity( Entity entity )
 		}
 	}
 
+	protected void Clear()
+	{
+		InteractionList.Clear();
+		Options.DeleteChildren( true );
+	}
+
 	public void CreatePanel()
 	{
+		int i = 0;
 		foreach ( Interaction interaction in InteractionList )
-			AddInteractionOption( interaction );
+		{
+			AddInteractionOption( interaction, i );
+			i++;
+		}
 	}
 
-	private void AddInteractionOption( Interaction interaction )
+	private void AddInteractionOption( Interaction interaction, int index )
 	{
 		if ( !interaction.CanResolve() )
 			return;
 
-		Label label = AddChild<Label>( "interaction-entry" );
-		label.Text = interaction.NiceName;
+		var label = Options.AddChild<InteractionEntry>( "interaction-entry" );
+		label.Name.Text = interaction.NiceName;
+		label.BindClass( "active-interaction", () => index == CurrentInteractionIndex );
 
-		label.AddEventListener( "onclick", () =>
-		{
-			interaction.ClientResolve();
+		UpdateArrows();
+	}
 
-			if ( interaction.ResolveOnServer )
-				Interaction.TryServerResolve( interaction.Owner.NetworkIdent, interaction.ID );
+	protected void Use()
+	{
+		var interaction = InteractionList[CurrentInteractionIndex];
 
-			foreach ( var child in Children )
-			{
-				if ( child == NameLabel )
-					continue;
+		interaction.ClientResolve();
 
-				child.Delete();
-			}
+		if ( interaction.ResolveOnServer )
+			Interaction.TryServerResolve( interaction.Owner.NetworkIdent, interaction.ID );
 
-			InteractionList.Clear();
+		Clear();
+	}
 
-		} );
+	protected void DoMouseWheelInput( int delta )
+	{
+		var isAscending = delta == 1;
+		var length = InteractionList.Count();
+
+		CurrentInteractionIndex += isAscending ? -1 : 1;
+		CurrentInteractionIndex = (CurrentInteractionIndex + length) % length;
+
+		UpdateArrows();
 	}
 
-	public override void Tick()
+	protected void UpdateArrows()
 	{
-		SetClass( "hide", InteractionList.Count == 0 );
+		UpArrow.SetClass( "active", ShouldShowArrow( false ) );
+		DownArrow.SetClass( "active", ShouldShowArrow( true ) );
 
-		if ( InteractionList.Count > 0 )
-			return;
+		UpArrow.SetClass( "hide", ShouldHideArrows() );
+		DownArrow.SetClass( "hide", ShouldHideArrows() );
+	}
+
+	[Event.BuildInput]
+	protected void BuildInput( InputBuilder input )
+	{
+		SetClass( "hide", InteractionList.Count == 0 );
 
 		var player = Local.Pawn;
 		var tr = Trace.Ray( player.EyePosition, player.EyePosition + player.EyeRotation.Forward * 200 )
 			.Ignore( player )
 			.Run();
 
-		if ( Debug.Enabled )
-			DebugOverlay.TraceResult( tr );
-
 		if ( tr.Hit && tr.Entity is Entity entity )
 		{
-			if ( Input.Pressed( InputButton.Use ) )
-			{
-				if ( entity is not IInteractable )
-					return;
+			if ( entity is not IInteractable )
+				return;
 
-				SetEntity( entity );
-			}
+			SetEntity( entity );
+		}
+		else
+		{
+			CurrentEntity = null;
+			Clear();
 
+			return;
 		}
+
+		if ( input.MouseWheel != 0f )
+			DoMouseWheelInput( input.MouseWheel );
+
+		if ( input.Pressed( InputButton.Use ) )
+			Use();
```

Comment:
Pressing or holding "Use" after you select an InteractionEntry is going to result in an ArgumentOutOfRange Exception here.

Reply:
I'll fix these tomorrow